File: proof_of_concept/computer/context_mapper.py
#!/home/fmeslet/anaconda3/bin/python3.8
#-*-coding: utf-8 -*-

import numpy as np

# Prime number to compute hash
A = 54059 # a prime
B = 76963 # another prime
C = 86969 # yet another prime
FIRSTH = 37 # also prime

class ContextMapper():

    def __init__(
        self, size_context, 
        size_element):

        self.size_context = size_context    
        self.size_element = size_element

        self.size_context_ravel = self.size_context*\
                self.size_element

        self.data_context = {}
        self.context_size = {}


    # Same implementation as C++ Arduino
    #def compute_hash(self, key_value):
    #    a = 63689
    #    b = 378551
    #    h = 0

    #    for i in range(len(key_value)):
    #        h = h * a + key_value[i])
    #        a = a * b

    #    return h


    def compute_hash(
        self, key_array):

        h = hash(''.join(
            [str(k) for k in key_array])) % 256

        return h


    def update_key(self, 
                  key_value, 
                  data):

        # Extract array with key
        context_data = self.data_context[
                key_value];


        # Shift context data
        for i in range(self.size_context-1):
            for j in range(self.size_element): 
              context_data[(i*self.size_element) + j] = \
                    context_data[(i+1)*self.size_element+j]

        # Update last values of context
        for j in range(len(data)): 
            context_data[
                (self.size_context-1)*self.size_element + j] = data[j]

        # Fill the other part with 0
        for j in range(len(data), self.size_element): 
            context_data[(self.size_context-1)*self.size_element+j] = 0


        # Add to map
        self.data_context[key_value] = context_data
        self.context_size[key_value] = min(
            self.context_size[key_value]+1, 
            self.size_context)

        print("[DEBUG][update_key] key_value: ", key_value)
        print("[DEBUG][update_key] context_data: ", context_data)
  

    def update(self, 
               key_array,
               data): # Check if array is well sized

        # Compute key
        key_value = self.compute_hash(
            key_array)

        # Check if the key exist
        if (key_value in self.data_context):
            #pass
            self.update_key(
                key_value, data)
        else:
            self.create_key(key_value)


    def create_key(
        self, key_value):

        # Init context to zero
        context_data = np.zeros(
            (self.size_context*self.size_element))

        # Add to map
        self.data_context[key_value] = context_data
        self.context_size[key_value] = 0


    def get_data_context(
        self, key_array):

      # Compute key
      key_value = self.compute_hash(
        key_array)

      # Check if the key exist
      if (key_value in self.data_context):
        #print("[DEBUG][ContextMapper::getDataContext] keyValue %d exist !\n", keyValue);
        return self.data_context[key_value]
      else:
        #print("[DEBUG][ContextMapper::getDataContext] keyValue %d not EXIST !\n", keyValue);
        self.create_key(key_value)
        return self.data_context[key_value]


    def get_context_size(
        self, key_array):

        # Compute key
        key_value = self.compute_hash(
            key_array)

        if (key_value in self.context_size):
            return self.context_size[key_value]
        else:
            self.context_size[key_value] = 0
            return self.context_size[key_value]

    def get_size_element(self):
        return self.size_element

    def get_size_element(self):
        return self.size_context
        

File: proof_of_concept/iot/Sender_performance/context_mapper.cpp
#include "context_mapper.h"


/*
 * CONSTRUCTOR
 */


ContextMapper::ContextMapper(
  int sizeC, int sizeE):
  sizeContext(sizeC),
  sizeElement(sizeE)
{  
  sizeContextRavel = sizeContext*sizeElement;
};


/*
 * OTHERS
 */


// For more information: https://stackoverflow.com/questions/8317508/hash-function-for-a-string
// https://www.partow.net/programming/hashfunctions/
// https://stackoverflow.com/questions/6136964/rs-hashing-program
//  RS Hash Function
int ContextMapper::computeHash(uint8_t* keyArray,
                               int sizeKeyArray)
{
   unsigned int b    = 378551;
   unsigned int a    = 63689;
   unsigned int hash = 0;
   unsigned int i    = 0;

   for (i = 0; i < sizeKeyArray; ++keyArray, ++i)
   {
      hash = hash * a + (*keyArray);
      a    = a * b;
   }

   return hash;
}


void ContextMapper::concatArray(
            uint8_t* arrayA,
            int sizeArrayA,
            uint8_t* arrayB,
            int sizeArrayB,
            uint8_t* arrayConcat)
{
  
  int sizeArrayConcat = sizeArrayA + sizeArrayB;
  
  // For on array A
  for (int i=0; i<sizeArrayA; i++)
  {
    arrayConcat[i] = arrayA[i];
  }

  // For on array B
  for (int i=sizeArrayA; i<sizeArrayConcat; i++)
  {
    arrayConcat[i] = arrayB[i-sizeArrayA];
  }
}



/*
 * Examples:
 * [0, 1, 2, 3, 4, ...]
 * The last one is the 4. The shift produce:
 * [1, 2, 3, 4, 5, ...]
 */

void ContextMapper::updateKey(
  int keyValue, 
  uint8_t* data,
  int sizeData)
{
  // Extract array with key
  uint8_t* contextData = dataContext[
        keyValue];

  // Shift context data
  for (int i=0; i<sizeContext-1; i++) { 
    for (int j=0; j<sizeElement; j++) { 
      contextData[(i*sizeElement) + j] = contextData[(i+1)*sizeElement+j]; 
    }
  }

  // Update last values of context
  for (int j=0; j<sizeData; j++) { 
      contextData[(sizeContext-1)*sizeElement + j] = data[j]; 
  }

  // Fill the other part with 0
  for (int j=sizeData; j<sizeElement; j++) { 
      contextData[(sizeContext-1)*sizeElement+j] = 0; 
  }

  // Add to map
  dataContext[keyValue] = contextData;
  contextSize[keyValue] = min(
    contextSize[keyValue]+1, sizeContext);
  
}


void ContextMapper::update(uint8_t* keyArray,
                      int sizeKeyArray,
                      uint8_t* data,
                      int sizeData) // Check if array is well sized
{
  // Compute key
  int keyValue = this->computeHash(
    keyArray, sizeKeyArray);
  printf("[DEBUG][ContextMapper::update] keyValue : %d \n", keyValue);
  
  // Check if the key exist
  if (dataContext.find(
    keyValue) != dataContext.end())
  {
    this->updateKey(
      keyValue, data, sizeData);
  }
  else
  {
    printf("[DEBUG][ContextMapper::update] keyValue NOT exist \n");
    this->createKey(keyValue);
    this->updateKey( // TO REMOVE ?
      keyValue, data, sizeData);
  }
  
}


void ContextMapper::createKey(
  int keyValue)
{
  // Init context to zero
  uint8_t* contextData = new uint8_t[
        sizeContext*sizeElement];
        
  for (int i=0; i<sizeContext*sizeElement; i++) { 
      contextData[i] = 0; 
  }

  // Add to map
  dataContext[keyValue] = contextData;
  contextSize[keyValue] = 0;

}



/*
 * GETTER
 */



uint8_t* ContextMapper::getDataContext(
  uint8_t* keyArray, int sizeKeyArray)
{
  // Compute key
  int keyValue = this->computeHash(
    keyArray, sizeKeyArray);

  // Check if the key exist
  if (dataContext.find(
    keyValue) != dataContext.end())
  {
    return dataContext[keyValue];
  }
  else
  {
    this->createKey(keyValue);
    return dataContext[keyValue];
  }
}


int ContextMapper::getContextSize(
      uint8_t* keyArray, 
      int sizeKeyArray)
{
  // Compute key
  int keyValue = this->computeHash(
    keyArray, sizeKeyArray);

  // Check if the key exist
  if (contextSize.find(
    keyValue) != contextSize.end())
  {
    contextSize[keyValue]; // return 0;
  }
  else
  {
    contextSize[keyValue] = 0;
  }

  return contextSize[keyValue];
}


int ContextMapper::getSizeElement()
{
  return sizeElement;
}


int ContextMapper::getSizeContext()
{
  return sizeContext;
}
